File: alerts.py
#!/usr/bin/env python3
"""Short-lived alert preparation; playback stays owned by Quickshell."""
import argparse
import fcntl
import html
import json
import math
import os
from pathlib import Path
import struct
import subprocess
import tempfile
import time
import wave

CACHE = Path(os.environ.get("XDG_CACHE_HOME", str(Path.home() / ".cache"))) / "omarchy-awqat" / "audio"
STATE = Path(os.environ.get("XDG_STATE_HOME", str(Path.home() / ".local/state"))) / "omarchy-awqat"
ADHANS = {"adhan-nafees": "https://cdn.aladhan.com/audio/adhans/a1.mp3",
          "adhan-alafasy": "https://cdn.aladhan.com/audio/adhans/a9.mp3"}
PRAYERS = {"Fajr", "Dhuhr", "Asr", "Maghrib", "Isha"}


def claim_event(event, now=None):
    """Persist before delivery, so reloads and multiple monitors cannot repeat it."""
    now = time.time() if now is None else now
    epoch = float(event.get("epoch", 0))
    if event.get("name") not in PRAYERS or not math.isfinite(epoch) or not 0 <= now - epoch <= 90:
        return False
    STATE.mkdir(parents=True, exist_ok=True, mode=0o700)
    STATE.chmod(0o700)
    with (STATE / "alerts.lock").open("a") as lock:
        deadline = time.monotonic() + 3
        while True:
            try:
                fcntl.flock(lock, fcntl.LOCK_EX | fcntl.LOCK_NB)
                break
            except BlockingIOError:
                if time.monotonic() >= deadline:
                    raise TimeoutError("Another alert is still being prepared. Please retry.")
                time.sleep(0.05)
        ledger_path = STATE / "delivered.json"
        try:
            if ledger_path.stat().st_size > 16384:
                raise ValueError("Oversized delivery ledger")
            ledger = json.loads(ledger_path.read_text())
            if not isinstance(ledger, dict):
                ledger = {}
        except (OSError, ValueError, RecursionError):
            ledger = {}
        ledger = {k: v for k, v in ledger.items() if isinstance(v, (float, int)) and math.isfinite(v) and 0 <= now - v < 3 * 86400}
        key = event["name"] + ":" + str(int(epoch))
        if key in ledger:
            return False
        ledger[key] = now
        ledger = dict(sorted(ledger.items(), key=lambda pair: pair[1])[-32:])
        temp = None
        try:
            with tempfile.NamedTemporaryFile(mode="w", dir=STATE, delete=False) as f:
                temp = f.name
                json.dump(ledger, f)
            os.replace(temp, ledger_path)
        finally:
            if temp:
                Path(temp).unlink(missing_ok=True)
        return True


def make_tone(path, kind):
    rate = 16000
    seconds = 2.2 if kind == "bell" else 1.8
    frames = bytearray()
    for i in range(int(rate * seconds)):
        t = i / rate
        value = 0
        notes = [(0, 660), (0.5, 880)] if kind == "chime" else [(0, 523.25)]
        for start, hz in notes:
            elapsed = t - start
            if elapsed < 0:
                continue
            envelope = min(1, elapsed / 0.015) * math.exp(-elapsed * 3.1)
            value += envelope * (math.sin(2 * math.pi * hz * elapsed) + 0.22 * math.sin(2 * math.pi * hz * 2.01 * elapsed))
        fade = min(1, (seconds - t) / 0.1)
        frames.extend(struct.pack("<h", int(max(-1, min(1, value * 0.32 * fade)) * 32767)))
    with wave.open(str(path), "wb") as output:
        output.setnchannels(1)
        output.setsampwidth(2)
        output.setframerate(rate)
        output.writeframes(frames)


def prepare_sound(settings):
    sound = settings.get("sound", "none")
    if sound == "none":
        return ""
    if sound == "custom":
        path = Path(str(settings.get("audioFile", ""))).expanduser()
        if not path.is_file() or path.suffix.lower() not in {".mp3", ".wav", ".ogg", ".flac", ".m4a", ".opus"}:
            raise ValueError("Choose an existing MP3, WAV, OGG, FLAC, M4A, or Opus audio file.")
        return str(path.resolve())
    if sound not in ADHANS and sound not in {"chime", "bell"}:
        raise ValueError("Unknown alert sound.")
    CACHE.mkdir(parents=True, exist_ok=True, mode=0o700)
    CACHE.chmod(0o700)
    path = CACHE / (sound + (".mp3" if sound in ADHANS else ".wav"))
    if path.is_file() and 1000 < path.stat().st_size <= 8388608:
        return str(path)
    with tempfile.NamedTemporaryFile(dir=CACHE, prefix=".audio-", delete=False) as f:
        temp = Path(f.name)
    try:
        if sound in ADHANS:
            result = subprocess.run(["curl", "-q", "-fsSL", "--max-redirs", "0", "--connect-timeout", "5", "--max-time", "25",
                "--max-filesize", "8388608", "--proto", "=https", "--proto-redir", "=https",
                "--output", str(temp), ADHANS[sound]], capture_output=True, timeout=28)
            with temp.open("rb") as f:
                header = f.read(3)
            if result.returncode or not 1000 <= temp.stat().st_size <= 8388608 or not (header == b"ID3" or header[:1] == b"\xff"):
                raise ValueError("Could not download the adhan. Check your connection and try Preview again.")
        else:
            make_tone(temp, sound)
        os.replace(temp, path)
    finally:
        temp.unlink(missing_ok=True)
    return str(path)


def deliver(settings, event):
    notifications = settings.get("notifications", False) is True
    sound_on = settings.get("sound", "none") != "none"
    if not notifications and not sound_on:
        return {"ok": True, "skipped": True}
    if not claim_event(event):
        return {"ok": True, "skipped": True}
    warning = ""
    if notifications:
        try:
            title = event["name"] + " · " + event.get("arabic", "")
            body = "It’s time for " + event["name"] + "." + ("\n" + str(event.get("city", "")) if event.get("city") else "")
            subprocess.run(["notify-send", "--app-name=Awqat", "--icon=appointment-soon",
                "--urgency=normal", "--expire-time=15000", "--", title, html.escape(body)], check=True, timeout=5)
        except (OSError, subprocess.SubprocessError):
            warning = "The desktop notification could not be delivered."
    try:
        audio = prepare_sound(settings) if sound_on else ""
    except Exception as error:
        return {"ok": False, "error": str(error)}
    # Never start audio late because an on-demand download took too long.
    if time.time() - float(event["epoch"]) > 90:
        audio = ""
    return {"ok": True, "file": audio, "warning": warning}


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--settings", default="{}")
    parser.add_argument("--event")
    parser.add_argument("--test-notification", action="store_true")
    args = parser.parse_args()
    try:
        settings = json.loads(args.settings)
        if not isinstance(settings, dict):
            raise ValueError("Settings must be a JSON object.")
        if args.test_notification:
            subprocess.run(["notify-send", "--app-name=Awqat", "--icon=appointment-soon", "--expire-time=6000",
                            "--", "Awqat · Test notification", "Prayer-time notifications are ready."], check=True, timeout=5)
            result = {"ok": True, "file": ""}
        else:
            result = deliver(settings, json.loads(args.event)) if args.event else {"ok": True, "file": prepare_sound(settings)}
    except Exception as error:
        result = {"ok": False, "error": str(error)}
    print(json.dumps(result, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: src/alerts.rs
use crate::{
    Result, Settings,
    net::{AUDIO_LIMIT, Network},
    storage,
};
use serde::Deserialize;
use serde_json::{Value, json};
use std::{
    collections::BTreeMap,
    fs,
    io::Read,
    path::Path,
    process::{Command, Stdio},
    thread,
    time::{Duration, Instant},
};

#[derive(Debug, Deserialize)]
pub struct Event {
    pub name: String,
    pub epoch: f64,
    #[serde(default)]
    pub arabic: String,
    #[serde(default)]
    pub city: String,
}

pub fn claim_event(event: &Event, state: &Path, now: f64) -> Result<bool> {
    if !["Fajr", "Dhuhr", "Asr", "Maghrib", "Isha"].contains(&event.name.as_str())
        || !event.epoch.is_finite()
        || !(0.0..=90.0).contains(&(now - event.epoch))
    {
        return Ok(false);
    }
    let _lock = storage::lock(state, "alerts.lock", 3)?;
    let path = state.join("delivered.json");
    let raw: Value = storage::read_bounded(&path, 16384)
        .ok()
        .and_then(|bytes| serde_json::from_slice(&bytes).ok())
        .unwrap_or_default();
    let mut ledger: BTreeMap<String, f64> = raw
        .as_object()
        .into_iter()
        .flatten()
        .filter_map(|(key, value)| value.as_f64().map(|value| (key.clone(), value)))
        .collect();
    ledger.retain(|_, value| value.is_finite() && (0.0..3.0 * 86400.0).contains(&(now - *value)));
    let key = format!("{}:{}", event.name, event.epoch as i64);
    if ledger.contains_key(&key) {
        return Ok(false);
    }
    ledger.insert(key, now);
    while ledger.len() > 32 {
        let oldest = ledger
            .iter()
            .min_by(|a, b| a.1.total_cmp(b.1))
            .map(|(k, _)| k.clone())
            .unwrap();
        ledger.remove(&oldest);
    }
    storage::atomic_write(&path, &serde_json::to_vec(&ledger)?)?;
    Ok(true)
}

pub fn tone(kind: &str) -> Vec<u8> {
    let rate: u32 = 16000;
    let seconds: f64 = if kind == "bell" { 2.2 } else { 1.8 };
    let count = (seconds * rate as f64) as u32;
    let mut output = Vec::with_capacity(44 + count as usize * 2);
    output.extend_from_slice(b"RIFF");
    output.extend_from_slice(&(36 + count * 2).to_le_bytes());
    output.extend_from_slice(b"WAVEfmt \x10\0\0\0\x01\0\x01\0");
    output.extend_from_slice(&rate.to_le_bytes());
    output.extend_from_slice(&(rate * 2).to_le_bytes());
    output.extend_from_slice(b"\x02\0\x10\0data");
    output.extend_from_slice(&(count * 2).to_le_bytes());
    let notes: &[(f64, f64)] = if kind == "bell" {
        &[(0.0, 523.25)]
    } else {
        &[(0.0, 660.0), (0.5, 880.0)]
    };
    for i in 0..count {
        let t = i as f64 / rate as f64;
        let mut value = 0.0;
        for (start, hz) in notes {
            let elapsed = t - start;
            if elapsed < 0.0 {
                continue;
            }
            let envelope = (elapsed / 0.015).min(1.0) * (-elapsed * 3.1).exp();
            let phase = std::f64::consts::TAU * hz * elapsed;
            value += envelope * (phase.sin() + 0.22 * (phase * 2.01).sin());
        }
        let sample = (value * 0.32 * ((seconds - t) / 0.1).min(1.0)).clamp(-1.0, 1.0);
        output.extend_from_slice(&((sample * 32767.0) as i16).to_le_bytes());
    }
    output
}
fn valid_audio_header(bytes: &[u8], mp3: bool) -> bool {
    if mp3 {
        bytes.starts_with(b"ID3")
            || (bytes.len() >= 2 && bytes[0] == 0xff && bytes[1] & 0xe0 == 0xe0)
    } else {
        bytes.starts_with(b"RIFF") && bytes.get(8..12) == Some(b"WAVE")
    }
}
pub fn prepare_sound(settings: &Settings, cache: &Path, network: &impl Network) -> Result<String> {
    let sound = settings.sound.as_str();
    if sound == "none" {
        return Ok(String::new());
    }
    if sound == "custom" {
        let expanded = if let Some(rest) = settings.audio_file.strip_prefix("~/") {
            std::path::PathBuf::from(std::env::var_os("HOME").ok_or("HOME is not set")?).join(rest)
        } else {
            settings.audio_file.clone().into()
        };
        let extension = expanded
            .extension()
            .and_then(|x| x.to_str())
            .unwrap_or("")
            .to_lowercase();
        if !expanded.is_file()
            || !["mp3", "wav", "ogg", "flac", "m4a", "opus"].contains(&extension.as_str())
        {
            return Err("Choose an existing MP3, WAV, OGG, FLAC, M4A, or Opus audio file".into());
        }
        return Ok(expanded.canonicalize()?.to_string_lossy().into_owned());
    }
    let url = match sound {
        "adhan-nafees" => Some("https://cdn.aladhan.com/audio/adhans/a1.mp3"),
        "adhan-alafasy" => Some("https://cdn.aladhan.com/audio/adhans/a9.mp3"),
        "chime" | "bell" => None,
        _ => return Err("Unknown alert sound".into()),
    };
    storage::private_dir(cache)?;
    let path = cache.join(format!(
        "{sound}.{}",
        if url.is_some() { "mp3" } else { "wav" }
    ));
    if let Ok(mut file) = fs::File::open(&path) {
        let meta = file.metadata()?;
        let mut header = [0; 12];
        if meta.is_file()
            && (1001..=AUDIO_LIMIT as u64).contains(&meta.len())
            && file.read_exact(&mut header).is_ok()
            && valid_audio_header(&header, url.is_some())
        {
            return Ok(path.to_string_lossy().into_owned());
        }
    }
    let bytes = if let Some(url) = url {
        network.get(url, AUDIO_LIMIT, 25)?
    } else {
        tone(sound)
    };
    if !(1000..=AUDIO_LIMIT).contains(&bytes.len()) || !valid_audio_header(&bytes, url.is_some()) {
        return Err("Could not download a valid adhan. Please retry.".into());
    }
    storage::atomic_write(&path, &bytes)?;
    Ok(path.to_string_lossy().into_owned())
}

pub fn escape(text: &str) -> String {
    text.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#x27;")
}
fn notify(title: &str, body: &str) -> Result<()> {
    let mut child = Command::new("notify-send")
        .args([
            "--app-name=Awqat",
            "--icon=appointment-soon",
            "--urgency=normal",
            "--expire-time=15000",
            "--",
            title,
            &escape(body),
        ])
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()?;
    let deadline = Instant::now() + Duration::from_secs(5);
    loop {
        match child.try_wait() {
            Ok(Some(status)) => {
                return if status.success() {
                    Ok(())
                } else {
                    Err("The desktop notification could not be delivered".into())
                };
            }
            Ok(None) if Instant::now() < deadline => thread::sleep(Duration::from_millis(20)),
            result => {
                let _ = child.kill();
                let _ = child.wait();
                return Err(match result {
                    Err(e) => e.into(),
                    _ => "Desktop notification timed out".into(),
                });
            }
        }
    }
}
pub fn test_notification() -> Result<()> {
    notify(
        "Awqat · Test notification",
        "Prayer-time notifications are ready.",
    )
}
pub fn deliver(
    settings: &Settings,
    event: &Event,
    cache: &Path,
    state: &Path,
    network: &impl Network,
) -> Result<Value> {
    if !settings.notifications && settings.sound == "none"
        || !claim_event(event, state, crate::now())?
    {
        return Ok(json!({"ok":true,"skipped":true}));
    }
    let mut warning = String::new();
    if settings.notifications {
        let title = format!("{} · {}", event.name, event.arabic);
        let body = format!(
            "It’s time for {}.{}",
            event.name,
            if event.city.is_empty() {
                String::new()
            } else {
                format!("\n{}", event.city)
            }
        );
        if notify(&title, &body).is_err() {
            warning = "The desktop notification could not be delivered.".into();
        }
    }
    let file = prepare_sound(settings, cache, network)?;
    Ok(
        json!({"ok":true,"file":if crate::now()-event.epoch>90.0 {String::new()} else {file},"warning":warning}),
    )
}
